packages/fxa-auth-server/lib/oauthdb/index.js

```diff
@@ -99,7 +105,56 @@ module.exports = (log, config) => {
 
     async grantTokensFromAuthorizationCode(oauthParams) {
       try {
-        return await api.grantTokensFromAuthorizationCode(oauthParams);
+        const codeResponse = await api.grantTokensFromAuthorizationCode(
+          oauthParams
+        );
+        const scopeSet = ScopeSet.fromString(codeResponse.scope);
+
+        if (scopeSet.contains(OAUTH_SCOPE_SESSION_TOKEN)) {
+          const origSessionToken = await db.sessionToken(
+            codeResponse.session_token
+          );
+          // Copy verification state of the token, but generate
+          // independent verification codes.
+          let newVerificationState = {};
+          if (origSessionToken.tokenVerificationId) {
+            newVerificationState.tokenVerificationId = random.hex(
+              origSessionToken.tokenVerificationId.length / 2
+            );
+          }
+
+          if (origSessionToken.tokenVerificationCode) {
+            // Using expiresAt=0 here prevents the new token from being verified via email code.
+            // That's OK, because we don't send them a new email with the new verification code
+            // unless they explicitly ask us to resend it, and resend only handles email links
+            // rather than email codes.
```

Comment:
Lol, yeah, I figure if you have to copy-paste an explanatory comment like this, it's probably a good sign that you should be extracting a shared helper function 😆